opencog/cython/PythonEval.cc

```diff
@@ -715,35 +716,71 @@ int PythonEval::argument_count(PyObject* pyFunction)
 }
 
 /**
- * Get the Python module and stripped function name given the identifer of
- * the form 'module.function'.
+ * Find the Python object by its name in the given module'.
  */
-PyObject* PythonEval::module_for_function(const std::string& moduleFunction,
-                                          std::string& functionName)
+PyObject* PythonEval::find_object(const PyObject* pyModule,
+                                  const std::string& objectName)
 {
+    PyObject* pyDict = PyModule_GetDict(_pyRootModule);
+    return PyDict_GetItemString(pyDict, objectName.c_str());
+}
+
+/**
+ * Get the Python module and/or object and stripped function name given the identifer of
+ * the form '[module.][object.[attribute.]*]function'.
+ */
+void PythonEval::module_for_function(const std::string& moduleFunction,
+                                     PyObject*& pyModule,
+                                     PyObject*& pyObject,
+                                     std::string& functionName)
+{
+    pyModule = _pyRootModule;
+    pyObject = nullptr;
+    functionName = moduleFunction;
     // Get the correct module and extract the function name.
     int index = moduleFunction.find_first_of('.');
     if (0 < index) {
         std::string moduleName = moduleFunction.substr(0, index);
-        PyObject* pyModule = _modules[moduleName];
-
-        // If not found, try loading it.
+        PyObject* pyModuleTmp = _modules[moduleName];
+        // If not found, first check that it is not an object.
+        // Then try loading it.
         // We have to guess, if its a single file, or an entire
         // directory with an __init__.py file in it ...
-        if (nullptr == pyModule) {
+        if (nullptr == pyModuleTmp &&
+            nullptr == find_object(pyModule, moduleName))
+        {
             add_modules_from_path(moduleName);
             add_modules_from_path(moduleName + ".py");
-            pyModule = _modules[moduleName];
+            pyModuleTmp = _modules[moduleName];
         }
-
-        // If found, we are done.
-        if (pyModule) {
+        // If found, set new module and truncate the function name
+        if (pyModuleTmp) {
+            pyModule = pyModuleTmp;
             functionName = moduleFunction.substr(index+1);
-            return pyModule;
         }
     }
-    functionName = moduleFunction;
-    return _pyRootModule;
+    // Iteratively check for objects in the selected (either root or loaded) module
+    index = functionName.find_first_of('.');
+    while (0 < index) {
+        std::string objectName = functionName.substr(0, index);
+        // If there is no object yet, find it in Module
+        // Else find it as Attr in Object
+        pyObject = nullptr == pyObject ?
+            find_object(pyModule, objectName) :
+            PyObject_GetAttrString(pyObject, objectName.c_str());
```

Comment:
```PyObject_GetAttrString()``` returns new reference, so you need to do ```Py_DECREF()``` for returned object after it is not needed. And ```find_object()``` returns result of ```PyDict_GetItemString()``` which is borrowed reference. So reference counter is different for first and second option.

Reply:
As an idea about managing memory here and below: you can create some small class which encapsulates all details which are required to call a function and use it instead of triple pyModule, pyObject, functionname. Then ```module_for_function``` will initialize this object and you can calculate and return number of arguments from it and so on.

Simple way is to wrap ```find_object``` by ```Py_INCREF``` or ```Py_XINCREF```